src/jsx.ts

```diff
@@ -0,0 +1,39 @@
+import {vnode, VNode, VNodeData} from './vnode';
+import {h, VNodeChildren} from './h';
+
+export type FunctionComponent = (props: {[prop: string]: any} | null, children?: VNode[]) => VNode;
+
+function flattenAndFilter(children: VNodeChildren[], flattened: VNode[]): VNode[] {
+  for (const child of children) {
+    // filter out falsey children, except 0 since zero can be a valid value e.g inside a chart
+    if (child !== undefined && child !== null && child !== false &&  child !== '') {
+      if (Array.isArray(child)) {
+        flattenAndFilter(child, flattened);
+      } else if (typeof child === 'string' || typeof child === 'number' || typeof child === 'boolean') {
+        flattened.push(vnode(undefined, undefined, undefined, String(child), undefined));
+      } else{
+        flattened.push(child);
+      }
+    }
+  }
+  return flattened;
+}
+
+/**
+ * jsx/tsx compatible factory function
+ * see: https://www.typescriptlang.org/docs/handbook/jsx.html#factory-functions
+ */
+export function jsx(tag: string | FunctionComponent, data: VNodeData | null, ...children: VNodeChildren[]): VNode {
+  const flatChildren = flattenAndFilter(children, []);
+  if (typeof tag === 'function') {
+    // tag is a function component
+    return tag(data, flatChildren);
+  } else {
+    if (flatChildren.length == 1 && !flatChildren[0].sel && flatChildren[0].text) {
+      // only child is a simple text node, pass as text for a simpler vtree
+      return h(tag, data, flatChildren[0].text);
+    } else {
+      return h(tag, data, flatChildren);
+    }
```

Comment:
Would this mean the same if it were the following (same but without the `else`)?
```ts
return h(tag, data, flatChildren);
```

Reply:
Not sure if I understand, Since we have rest args `...children: VNodeChildren`, passing in a text node would always be encapsulated in an array[].

In terms of what snabbdom does, `h(tag, data, flatChildren)` would still work as h converts primitives to vnodes with text. The if just makes the vtree simpler.

Reply:
If all previous branches `return` then there's no need for this `return` to be in an `else`. It could be after the if-else structure. But that's subject to code style preferences. Something we could possibly lint for.

Reply:
Typescript compilation will fail if all branches don't return the same type. I am not sure what the proposed fix here is.

Reply:
I think I understand what you mean. Will change to keep you happy.

Reply:
tried the fix with return outermost, kinda reads funny to be honest. May be personal preference but I like all if branches to be accounted for and this explicitly accounts for them.